gitlab/resource_gitlab_branch.go

```diff
@@ -0,0 +1,181 @@
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"ref": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+				Default:  "master",
+			},
+			"web_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"default": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"can_push": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"merged": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"commit": {
+				Type:     schema.TypeList,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"author_email": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"author_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"authored_date": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"committed_date": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"committer_email": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"committer_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"short_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"title": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"message": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceGitlabBranchCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	name := d.Get("name").(string)
+	project := d.Get("project").(string)
+	ref := d.Get("ref").(string)
+	branchOptions := &gitlab.CreateBranchOptions{
+		Branch: &name, Ref: &ref,
+	}
+
+	log.Printf("[DEBUG] create gitlab branch %s for project %s with ref %s", name, project, ref)
+	branch, resp, err := client.Branches.CreateBranch(project, branchOptions)
+	if err != nil {
+		log.Printf("[DEBUG] failed to create gitlab branch %v response %v", branch, resp)
+		return err
+	}
+	return resourceGitlabBranchRead(d, meta)
+}
+
+func resourceGitlabBranchRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	name := d.Get("name").(string)
+	project := d.Get("project").(string)
+	ref := d.Get("ref").(string)
+	log.Printf("[DEBUG] read gitlab branch %s", name)
+	branch, resp, err := client.Branches.GetBranch(project, name)
```

Comment:
Please handle a 404 error here, and set the ID to `""` in that case. This ensures that the branch can be safely removed inside GitLab out-of-band. See https://learn.hashicorp.com/tutorials/terraform/provider-setup#implement-read